### PR TITLE
Runtime Versioning

### PR DESCRIPTION
**Pull request by jekloudaMSFT**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

This PR implements versioning for the runtime interface to unblock the adaptive card work. With these changes, we can make breaking changes to the runtime interface while still being maintaining compatibility with older versions of the runtime.

### Main changes in the PR:

1. Use the Runtime type alias to reference the latest runtime interface
2. Rename current runtime interface to IRuntimeV1
3. Implement fastForwardRuntime, a function that upgrades outdated runtime objects to use the latest interface
5. Add RUNTIME.md which describes when and how to update the runtime version
6. Add a GitHub Action that will leave a comment on PRs that touch the runtime.ts file to prompt the author to review RUNTIME.md

### Validation performed:

1. Implemented a version 0 runtime interface to test the version upgrade mechanism
2. GitHub Action tested by comments on this PR

### Unit Tests added:

Yes

### End-to-end tests added:

No

### Change file added:

> Ensure the change file meets the [formatting requirements](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball).

Yes

### Next/remaining steps:

This change will unblock the adaptive card implementation currently in PR

### CODE REVIEW

**Comment by jekloudaMSFT** [comment]
> This pull request contains changes to the runtime.ts file. If you, as the author of this PR, have made changes to the Runtime interface please review RUNTIME.md to determine if a new runtime version is required. Please reply to this comment stating what changes, if any, were made to the runtime object and whether a new runtime version was required.

Added runtime versioning. No updates to current runtime interface so no new version is required

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [inline packages/teams-js/src/public/runtime.ts:102]: Out of curiosity, what caused you to need to bump this from -1 to 1? Since it's uninitialized I assumed it should be set to some sort of error value

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [inline packages/teams-js/src/public/runtime.ts:168]: This should be hidden already since it isn't exported, but might be good to add the @hidden keyword just so no one accidentally exports it and starts writing dev documentation for it

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [inline packages/teams-js/src/public/runtime.ts:185]: Might be good to open a task on our backlog for handling truly breaking runtime changes at some point

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [inline packages/teams-js/src/public/runtime.ts:309]: Probably would be good to use applyRuntimeConfigLogger to write a log statement inside of this if statement as well, since mobile hosts that do this might cause weirdness that we need to debug in the future.

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [inline packages/teams-js/src/public/runtime.ts:18]: should this be renamed to LatestRuntime?

**Review by TrevorJoelHarris — DISMISSED** [review]
Left some questions, but looks good. If you need another review after responding to them let me know

File: packages/teams-js/src/public/runtime.ts
Comment (jekloudaMSFT) [reply packages/teams-js/src/public/runtime.ts:23]: This is something that came up when I was testing the upgrade mechanism. I wrote an IUpgrade that did a v0 to v1 transformation, and I needed an isV1 check so that the `upgrade` function could return a v1 object. I thought it might be good to just codify writing type checks as part of the upgrade process and then using them in this function, but maybe it would be better to leave that up to developers implementing upgrades.

Comment (jekloudaMSFT) [reply packages/teams-js/src/public/runtime.ts:23]: The other thing I was trying to do was tie the `latestRuntimeApiVersion` to the upgrade process. `isLatestRuntimeVersion` is used to verify that the runtime we have at the end of the upgrade chain is the version we want, so using `latestRuntimeApiVersion` here makes sure that number is always in sync with the Runtime interface, and then isRuntimeV1 would actually make sure the object is well-formed and has all required properties.

File: packages/teams-js/src/public/runtime.ts
Comment (jekloudaMSFT) [reply packages/teams-js/src/public/runtime.ts:102]: The version needs to be 1 to conform to the Runtime interface so it can be assigned to `runtime`. I looked into making `runtime` a union type with `UnitializedRuntime`, but we would have to type check `runtime` everywhere we use it. I like the idea of having an error value here, but we weren't actually checking for it anywhere, right? Maybe something like an `isInitialized`/`isUnintialized` flag would be helpful?

File: RUNTIME.md
Comment (jekloudaMSFT) [reply RUNTIME.md:35]: Good idea - I'll open a task for it

File: packages/teams-js/src/public/runtime.ts
Comment (jekloudaMSFT) [reply packages/teams-js/src/public/runtime.ts:18]: It could, but most code using the `runtime` object shouldn't be concerned about versioning, so I think it might be more helpful to think of this as "this is the `Runtime` interface" rather than leaking the versioning terminology into the rest of the codebase. What do you think?

File: packages/teams-js/src/public/runtime.ts
Comment (jekloudaMSFT) [reply packages/teams-js/src/public/runtime.ts]: I consolidated this to a single type check. Looks like we can do the version transformations without explicit type checks so I'll leave that up to whoever's implementing them. This makes the upgrade process a little more straightforward as well

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/public/runtime.ts:98]: Yes, I don't think we were checking it anywhere it was mostly just curiosity. Does that uninitializedRuntime object actually match what we expect from version 1? It looks like it doesn't have the expected capabilities in the supports object

File: packages/teams-js/src/public/runtime.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/public/runtime.ts:18]: okay, that's fine. It's internal, so easy to change if we change our minds

**Review by TrevorJoelHarris — APPROVED** [review]

File: packages/teams-js/src/public/runtime.ts
Comment (jekloudaMSFT) [reply packages/teams-js/src/public/runtime.ts]: Updated to use an `IUninitializedRuntime` interface for the uninitialized runtime

**Review by TrevorJoelHarris — APPROVED** [review]

**Review by jadahiya-MSFT — APPROVED** [review]

**Pull request merged**
